go/vt/vtgate/engine/hash_join.go

```diff
@@ -0,0 +1,257 @@
+			continue
+		}
+		hashcode, err := evalengine.NullsafeHashcode(joinVal, hj.Collation, hj.ComparisonType)
+		if err != nil {
+			return nil, err
+		}
+		lftRows := probeTable[hashcode]
+		for _, currentLHSRow := range lftRows {
+			lhsVal := currentLHSRow[hj.LHSKey]
+			// hash codes can give false positives, so we need to check with a real comparison as well
+			cmp, err := evalengine.NullsafeCompare(joinVal, lhsVal, collations.Unknown)
+			if err != nil {
+				return nil, err
+			}
+
+			if cmp == 0 {
+				// we have a match!
+				result.Rows = append(result.Rows, joinRows(currentLHSRow, currentRHSRow, hj.Cols))
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func (hj *HashJoin) buildProbeTable(lresult *sqltypes.Result) (map[evalengine.HashCode][]row, error) {
+	probeTable := map[evalengine.HashCode][]row{}
+	for _, current := range lresult.Rows {
+		joinVal := current[hj.LHSKey]
+		if joinVal.IsNull() {
+			continue
+		}
+		hashcode, err := evalengine.NullsafeHashcode(joinVal, hj.Collation, hj.ComparisonType)
+		if err != nil {
+			return nil, err
+		}
+		probeTable[hashcode] = append(probeTable[hashcode], current)
+	}
+	return probeTable, nil
+}
+
+// TryStreamExecute implements the Primitive interface
+func (hj *HashJoin) TryStreamExecute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, callback func(*sqltypes.Result) error) error {
+	// build the probe table from the LHS result
+	probeTable := map[evalengine.HashCode][]row{}
+	var lfields []*querypb.Field
+	err := vcursor.StreamExecutePrimitive(hj.Left, bindVars, wantfields, func(result *sqltypes.Result) error {
+		if len(lfields) == 0 && len(result.Fields) != 0 {
+			lfields = result.Fields
+		}
+		for _, current := range result.Rows {
+			joinVal := current[hj.LHSKey]
+			if joinVal.IsNull() {
+				continue
+			}
+			hashcode, err := evalengine.NullsafeHashcode(joinVal, hj.Collation, hj.ComparisonType)
+			if err != nil {
+				return err
+			}
+			probeTable[hashcode] = append(probeTable[hashcode], current)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return vcursor.StreamExecutePrimitive(hj.Right, bindVars, wantfields, func(result *sqltypes.Result) error {
+		// compare the results coming from the RHS with the probe-table
+		res := &sqltypes.Result{}
+		if len(result.Fields) != 0 {
+			res = &sqltypes.Result{
+				Fields: joinFields(lfields, result.Fields, hj.Cols),
+			}
+		}
+		for _, currentRHSRow := range result.Rows {
+			joinVal := currentRHSRow[hj.RHSKey]
+			if joinVal.IsNull() {
+				continue
+			}
+			hashcode, err := evalengine.NullsafeHashcode(joinVal, hj.Collation, hj.ComparisonType)
+			if err != nil {
+				return err
+			}
+			lftRows := probeTable[hashcode]
+			for _, currentLHSRow := range lftRows {
+				lhsVal := currentLHSRow[hj.LHSKey]
+				// hash codes can give false positives, so we need to check with a real comparison as well
+				cmp, err := evalengine.NullsafeCompare(joinVal, lhsVal, collations.Unknown)
```

Comment:
we can use column collation here instead of Unknown, WDYT?

Reply:
:facepalm: of course